packages/common/src/models/uri.ts

```diff
@@ -38,6 +38,25 @@ export function parseMidUri(
   }
 }
 
+export function bodyUri(cacheId: string, partId: string): URI {
+  const cId = idFromHeaderValue(cacheId)
+  const pId = idFromHeaderValue(partId)
+  return `body:${encodeURIComponent(cId)}/${encodeURIComponent(pId)}`
+}
+
+const bodyExp = /(body:)([^/]+)(?:\/(.+))?$/
```

Comment:
I think I told you that this expression would either match both the cache ID and part ID, or `matches` would be `null`. But I was mistaken. Because there is a question mark around one of the groups this expression will match if no part ID is given. Because this URI scheme is only useful when a part ID is given I suggest changing this regular expression.

The sub-expression `(?:\/(.+))?` optionally matches a slash and following text. The use of `?:` inside the outer pair of parenthesis makes it a non-capturing group. That is useful because it allows the entirety of the text captured by the group to be optional (due to the `?` at the end). Content matched by a non-capturing group will not be listen in the `matches` array - that array only includes "captured" groups. You can change this portion of the expression to be required by removing the last `?`. In that case it is not necessary to specify a group that includes the slash. You can change the expression to look like this:

```ts
const bodyExp = /body:([^/]+)\/(.+)$/
```

Notice that I also removed the parenthesis from around `body:` because we do not need to capture the scheme to examine its value. If you remove those parenthesis you need to change the indexes used to get the cache ID and part ID from the `matches` array.